plugins/cpp_reparse/service/src/astnodelocator.h

```diff
@@ -0,0 +1,88 @@
+#ifndef CC_SERVICE_CPPREPARSESERVICE_ASTNODELOCATOR_H
+#define CC_SERVICE_CPPREPARSESERVICE_ASTNODELOCATOR_H
+
+#include <clang/AST/ASTContext.h>
+
+#include <model/cppastnode.h>
+#include <model/cppastnode-odb.hxx>
+
+#include <util/logutil.h>
+
+#include <cppparser/filelocutil.h>
+
+namespace cc
+{
+namespace service
+{
+namespace reparse
+{
+
+/**
+ * This helper class provides a way to match a Clang AST Node (whichever actual
+ * type it might have) against a source location stored in the database.
+ */
+class ASTNodeLocator
+{
+public:
+  ASTNodeLocator(const clang::ASTContext& context_,
+                 const model::CppAstNodePtr dbNode_)
+    : _context(context_),
+      _locUtil(parser::FileLocUtil(context_.getSourceManager())),
+      _dbNode(dbNode_)
+  {}
+
+  /**
+   * Attempts to match the given Clang AST Node type (usually clang::Decl or
+   * clang::Stmt) against the CodeCompass C++ AST Node database row that the
+   * current class was instantiated with.
+   *
+   * Returns true if the location of the given node matches the "filter".
+   */
+  // Template method, because clang::Decl and clang::Stmt are two distinct
+  // hierarchies.
+  template <typename ClangASTNode>
+  bool matchNodeAgainstLocation(ClangASTNode* node_)
+  {
+    // TODO: What to do with CppAstNodes that are not in a source file? (generated things?)
+
+    if (node_)
+    {
+      // This is merely an optimisation so that no extra calculations are made.
+      // If the start line of the Decl does not match the filter, it can't be
+      // the ASTNode we are searching for.
+      clang::SourceLocation start = node_->getLocStart();
+      if (_context.getSourceManager().getPresumedLineNumber(start) !=
+          _dbNode->location.range.start.line)
+        return false;
+
+      clang::PresumedLoc presumedLoc =
+        _context.getSourceManager().getPresumedLoc(start);
```

Comment:
Isn't it slightly faster to query this presumed location instead of the presumed line number in the previous "if" statement? This way there was only one query for the location, and that contains the line number anyway.